core/arch/arm/mm/shmem.c

```diff
@@ -0,0 +1,587 @@
+{
+	unsigned int exceptions;
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	cpu_spin_lock(&spin_lock);
+	return exceptions;
+}
+
+static void shmem_unlock(unsigned int exceptions)
+{
+	cpu_spin_unlock(&spin_lock);
+	thread_unmask_exceptions(exceptions);
+}
+
+/**
+ * shmem_init() - Initialzie shared memory subsystem
+ */
+
+static TEE_Result shmem_init(void)
+{
+	vaddr_t pool_start;
+	vaddr_t pool_end;
+
+	core_mmu_get_mem_by_type(MEM_AREA_SHM_VASPACE, &pool_start, &pool_end);
+	if (pool_start == 0 || pool_end == 0)
+		panic("Can't find region for shmem pool");
+
+	if (!tee_mm_init(&tee_mm_shm, pool_start, pool_end, SMALL_PAGE_SHIFT,
+		    TEE_MM_POOL_NO_FLAGS))
+		panic("Could not create shmem pool");
+
+	SLIST_INIT(&mappings_list);
+	IMSG("Shared memory subsystem initialized: %lx, %lx\n",
+	     pool_start, pool_end);
+	return TEE_SUCCESS;
+}
+
+service_init(shmem_init);
+
+/**
+ * find_by_pa() - Finds mapping by physical address.
+ * @pa:		Physical address of buffer
+ *
+ * This function should be called while holding lock. It takes into account
+ * size of mapped buffer.
+ * Return:
+ *   NULL - if there are no such buffer
+ *   pointer to struct shmem_mapping with information about shared bufferx
+ */
+
+static struct shmem_mapping *find_by_pa(paddr_t pa)
+{
+	struct shmem_mapping *area = NULL;
+
+	SLIST_FOREACH(area, &mappings_list, next) {
+		/* Special case for first page */
+		if (pa >= area->pa &&
+		    pa < area->pages[0] + SMALL_PAGE_SIZE &&
+		    pa < area->pa + area->size)
+			return area;
+
+		/* Special case for last page */
+		if (area->pages_cnt > 1 &&
+		    pa >= area->pages[area->pages_cnt-1] &&
+		    pa < (area->pages[area->pages_cnt-1] +
+			  ((area->pa + area->size) & SMALL_PAGE_SIZE)))
+			return area;
+
+		/* All other cases are trivial */
+		for (int i = 1; i < area->pages_cnt-1; i++) {
+			if (pa >= area->pages[i] &&
+			    (pa < area->pages[i] + SMALL_PAGE_SIZE))
+				return area;
+		}
+	}
+	return NULL;
+}
+
+
+/**
+ * shmem_find_by_pa() - Finds mapping by physical address.
+ * @pa:		Physical address of buffer
+ *
+ * This function is exported, so it locks internal data and just
+ * calls for find_by_pa()
+ * Return:
+ *   NULL - if there are no such buffer
+ *   pointer to struct shmem_mapping with information about shared bufferx
+ */
+
+struct shmem_mapping *shmem_find_by_pa(paddr_t paddr)
+{
+	unsigned int exceptions = shmem_lock();
+	struct shmem_mapping *mapping = find_by_pa(paddr);
+
+	shmem_unlock(exceptions);
+	return mapping;
+}
+
+/**
+ * shmem_map_buffer() - map given buffer into OP-TEE virtual address space
+ * @pa:		Physical address of buffer beginning
+ * @size:		Size of buffer
+ * @pages:		Array of page addresses that hold buffer
+ * @pages_cnt:		Number of entries in @pages
+ * @shmem_mapping:	Pointer to new mapping data
+ *
+ * Return:
+ *   TEE_SUCCESS - on success
+ *   TEE_ERROR_XXX - on error
+ */
+
+int shmem_map_buffer(paddr_t pa, size_t size, paddr_t *pages,
```

Comment:
The page list dynamic mapping is a common function,  can we introduce a new function like map_page_list,  and the shmem_map_buffer is a wrapper function of the map_page_list here?